src/addonindex.cpp

```diff
@@ -0,0 +1,225 @@
+}
+
+/**
+ * @brief Gets the on disk addon list from the addons file, if any.
+ *
+ * @param addonsList Variable that will contain the index data if reading and
+ * validation of data succeeds.
+ * @return Whether or not the index file could be read / validated.
+ */
+bool AddonIndex::getOnDiskAddonsList(QList<AddonData>* addonsList) {
+  QByteArray index;
+  QByteArray indexSignature;
+  if (!read(index, indexSignature)) {
+    return false;
+  }
+
+  QJsonObject indexObj;
+  if (!validate(index, indexSignature, &indexObj)) {
+    logger.debug() << "Unable to validate the index";
+    return false;
+  }
+
+  *addonsList = extractAddonsFromIndex(indexObj);
+  return true;
+}
+
+/**
+ * @brief Reads the index files on disk, if any.
+ *
+ * @param index Variable that will contain the index data if
+ * reading succeeds.
+ * @param indexSignature Variable that will contain the index
+ * signature if reading succeeds.
+ * @returns Whether or not reading of both files was successfull.
+ */
+bool AddonIndex::read(QByteArray& index, QByteArray& indexSignature) {
+  return m_addonDirectory->readFile(ADDON_INDEX_FILENAME, &index) &&
+         m_addonDirectory->readFile(ADDON_INDEX_SIGNATURE_FILENAME,
+                                    &indexSignature);
+}
+
+/**
+ * @brief Writes new content to the on disk index files.
+ *
+ * Note: If the addons directory does not exist, this is no-op.
+ *
+ * @param index The contents to write to the index file.
+ * @param indexSignature The contents to write to the index signature file.
+ */
+void AddonIndex::write(const QByteArray& index,
+                       const QByteArray& indexSignature) {
+  m_addonDirectory->writeToFile(ADDON_INDEX_FILENAME, index);
+  m_addonDirectory->writeToFile(ADDON_INDEX_SIGNATURE_FILENAME, indexSignature);
+}
+
+/**
+ * @brief Updates the index on file in case provided index and indexSignature
+ * are valid and different from the currently stored ones.
+ *
+ * Note: If update is successful `indexUpdated` signal is emited. If nothing has
+ * changed, this signal is not emited.
+ *
+ * @param index The new index data to update.
+ * @param indexSignature The new index signature to update.
+ * @param ignoreUnchanged Whether or not to ignore the fact that the index has
+ * not changed. Defaults to `false`.
+ * @return Whether or not an update signal was emited.
+ */
+void AddonIndex::update(const QByteArray& index,
+                        const QByteArray& indexSignature) {
+  QByteArray currentIndex;
+  QByteArray currentIndexSignature;
+  if (read(currentIndex, currentIndexSignature) && currentIndex == index &&
+      currentIndexSignature == indexSignature) {
+    logger.debug() << "The index has not changed";
+    return;
+  }
+
+  QJsonObject indexObj;
+  if (!validate(index, indexSignature, &indexObj)) {
+    logger.debug() << "Unable to validate the index";
+    return;
+  }
+
+  // Question: Should we not emit the signal in case write is unsuccessfull?
```

Comment:
@bakulf WDYT?

Reply:
Yes, good point. we do something similar in case the validation fails.